java/core/src/java/org/apache/orc/OrcUtils.java

```diff
@@ -71,10 +79,23 @@ private static TypeDescription findColumn(String columnName,
                                             List<String> fieldNames,
                                             List<TypeDescription> fields) {
     int i = 0;
-    for(String fieldName: fieldNames) {
+    for (String fieldName: fieldNames) {
+      TypeDescription resultTypeDesc = fields.get(i);
+
       if (fieldName.equalsIgnoreCase(columnName)) {
-        return fields.get(i);
+        return resultTypeDesc;
+
       } else {
+        if (resultTypeDesc.getCategory() == TypeDescription.Category.STRUCT
+                && fieldName.equalsIgnoreCase(StringUtils.substringBefore(columnName, DOT))) {
```

Comment:
This should already work if Hive sends "row.<fieldname>" in selected columns?

Reply:
yes